sources/api_client.py

```diff
@@ -0,0 +1,217 @@
+
+    @next_reference.setter
+    def next_reference(self, value: Optional[str]):
+        self._next_reference = value
+        self._has_next_page = value is not None
+
+    @abstractmethod
+    def update_state(self, response: Response) -> None:
+        """Update the paginator state based on the response.
+
+        Args:
+            response (Response): The response object from the API.
+        """
+        ...
+
+    @abstractmethod
+    def prepare_next_request_args(
+        self, url: str, params: Optional[Dict[str, Any]], json: Optional[Dict[str, Any]]
+    ) -> Tuple[Optional[str], Optional[Dict[str, Any]], Optional[Dict[str, Any]]]:
+        """
+        Prepare the arguments for the next API request based on the current state of pagination.
+
+        Subclasses must implement this method to update the request arguments appropriately.
+
+        Args:
+            url (str): The original URL used in the current API request.
+            params (Optional[Dict[str, Any]]): The original query parameters used in the current API request.
+            json (Optional[Dict[str, Any]]): The original JSON body of the current API request.
+
+        Returns:
+            tuple: A tuple containing the updated URL, query parameters, and JSON body to be used
+                for the next API request. These values are used to progress through the paginated data.
+        """
+        ...
+
+    @abstractmethod
+    def extract_records(self, response: Response) -> Any:
+        """
+        Extract the records data from the response.
+
+        Args:
+            response (Response): The response object from the API.
+
+        Returns:
+            Any: The extracted records data.
+        """
+        ...
+
+
+class BaseNextUrlPaginator(BasePaginator):
+    def prepare_next_request_args(self, url, params, json):
+        return self._next_reference, params, json
+
+
+class HeaderLinkPaginator(BaseNextUrlPaginator):
+    """A paginator that uses the 'Link' header in HTTP responses
+    for pagination.
+
+    A good example of this is the GitHub API:
+        https://docs.github.com/en/rest/guides/traversing-with-pagination
+    """
+
+    def __init__(self, links_next_key: str = "next") -> None:
+        """
+        Args:
+            links_next_key (str, optional): The key (rel ) in the 'Link' header
+                that contains the next page URL. Defaults to 'next'.
+        """
+        super().__init__()
+        self.links_next_key = links_next_key
+
+    def update_state(self, response: Response) -> None:
+        self.next_reference = response.links.get(self.links_next_key, {}).get("url")
+
+    def extract_records(self, response: Response) -> Any:
+        return response.json()
+
+
+class JSONResponsePaginator(BaseNextUrlPaginator):
+    """A paginator that uses a specific key in the JSON response to find
+    the next page URL.
+    """
+
+    def __init__(self, next_key: str = "next", records_key: str = "results"):
```

Comment:
I think we should give option of json path here because I believe someone will come asking how can they extract next link from nested structure maybe this [jmespath-py](https://github.com/jmespath/jmespath.py) is good?